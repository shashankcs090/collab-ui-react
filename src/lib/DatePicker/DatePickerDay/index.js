import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@collab-ui/react';
import {
  getDate,
  getMonth,
  isDayDisabled,
  isSameDay,
  now,
} from '@collab-ui/react/DatePicker/date_utils';
import moment from 'moment';

class DatePickerDay extends React.Component {
  static displayName = 'DatePickerDay';

  handleClick = e => {
    const { handleDayClick } = this.context;
    const { day } = this.props;
    return (
      handleDayClick
      && handleDayClick(e, day)
    );
  };

  render() {
    const { day, month } = this.props;
    const { selected, focus } = this.context;

    const isOutsideMonth = month !== getMonth(day);
    const isSelected = isSameDay(day, selected);
    const isToday = isSameDay(day, now());
    const disabled = isDayDisabled(day, this.props);
    const hasFocus = isSameDay(day, focus);

    return (
      <Button
        circle
        size={28}
        disabled={disabled}
        className={
          'cui-datepicker__day' +
          `${(isSelected && ` cui-datepicker__day--selected`) || ''}` +
          `${(hasFocus && ` cui-datepicker__day--focus`) || ''}` +
          `${(isToday && ` cui-datepicker__day--today`) || ''}` +
          `${(isOutsideMonth && ` cui-datepicker__day--outside-month`) || ''}`
        }
        onClick={this.handleClick}
        ariaLabel={`${getDate(day)}`}
        aria-selected={isSelected}
        tabIndex={-1}
      >
        {getDate(day)}
      </Button>
    );
  }
}

DatePickerDay.contextTypes = {
  focus: PropTypes.instanceOf(moment),
  handleDayClick: PropTypes.func,
  selected: PropTypes.instanceOf(moment),
};

DatePickerDay.propTypes = {
  day: PropTypes.instanceOf(moment).isRequired,
  month: PropTypes.number.isRequired,
};

export default DatePickerDay;
