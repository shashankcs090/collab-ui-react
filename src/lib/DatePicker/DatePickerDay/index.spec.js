import React from 'react';
import { shallow, mount } from 'enzyme';
import DatePickerDay from '../DatePickerDay';
import moment from 'moment-timezone';

const TEST_DATE = '2018-04-21';

describe.only('tests for <DatePickerDay />', () => {
  let day;

  beforeAll(() => {
    // setting timezone to CST
    moment.tz.setDefault('Asia/Kolkata');
    day = moment(TEST_DATE);
  });

  afterAll(() => {
    moment.tz.setDefault();
  });

  it('should match selected day SnapShot', () => {
    const selected = moment(TEST_DATE);
    const container = shallow(
      <DatePickerDay day={day} month={3} selected={selected} />
    );
    expect(container).toMatchSnapshot();
  });

  describe('should apply the right modifier classes', () => {
    it('should apply a modifier, when the date does not belong to current month', () => {
      const container = mount(
        <DatePickerDay day={day} month={10} />
      );
      expect(container.find('.cui-button').hasClass('cui-datepicker__day--outside-month')).toEqual(true);
    });

    it('should apply a modifier when the date is selected', () => {
      const container = mount(
        <DatePickerDay day={day} month={4} />,
        {
          context: { selected: day }
        }
      );
      expect(container.find('.cui-button').hasClass('cui-datepicker__day--selected')).toEqual(true);
    });

    it('should apply a modifier when the date is today`s date', () => {
      const container = mount(
        <DatePickerDay day={moment()} month={4} />
      );
      expect(container.find('.cui-button').hasClass('cui-datepicker__day--today')).toEqual(true);
    });

    it('should apply a modifier when the date is in focus', () => {
      const container = mount(
        <DatePickerDay day={day} month={4} />,
        {
          context: { focus: day }
        }
      );
      expect(container.find('.cui-button').hasClass('cui-datepicker__day--focus')).toEqual(true);
    });
  });

  it('when Day the disabled, should set disabled prop on button', () => {
    const container = mount(
      <DatePickerDay day={day} month={4} filterDate={() => true} />
    );
    expect(container.find('.cui-button').props().disabled).toEqual(true);
  });

  it('onclick of Day should call the callback in context', () => {
    const onClickFn = jest.fn();
    const container = mount(
      <DatePickerDay day={day} month={4} focus={day} />,
      { context: { handleDayClick: onClickFn }}
    );
    container.find('.cui-button').simulate('click');
    expect(onClickFn).toHaveBeenCalled();
  });

  it('on press of enter/space key on Day, should call the callback in context', () => {
    const onClickFn = jest.fn();
    const container = mount(
      <DatePickerDay day={day} month={4} focus={day} />,
      { context: { handleDayClick: onClickFn }}
    );
    container.find('.cui-button').simulate('keyDown', { which: 13 });
    expect(onClickFn).toHaveBeenCalled();
    container.find('.cui-button').simulate('keyDown', { which: 32 });
    expect(onClickFn).toHaveBeenCalledTimes(2);
  });

});
