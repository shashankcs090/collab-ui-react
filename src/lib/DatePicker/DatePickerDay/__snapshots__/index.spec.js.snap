// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <DatePickerDay /> should match selected day SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DatePickerDay
    day={"2018-04-20T18:30:00.000Z"}
    month={3}
    selected={"2018-04-20T18:30:00.000Z"}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "active": false,
      "aria-selected": false,
      "ariaLabel": "21",
      "ariaLabelledBy": "",
      "children": 21,
      "circle": true,
      "className": "cui-datepicker__day",
      "color": "",
      "containerLarge": false,
      "disabled": false,
      "expand": false,
      "href": "",
      "index": null,
      "label": "",
      "large": false,
      "loading": false,
      "onClick": [Function],
      "size": 28,
      "style": Object {},
      "tabIndex": -1,
      "tag": "button",
      "type": "button",
    },
    "ref": null,
    "rendered": 21,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "active": false,
        "aria-selected": false,
        "ariaLabel": "21",
        "ariaLabelledBy": "",
        "children": 21,
        "circle": true,
        "className": "cui-datepicker__day",
        "color": "",
        "containerLarge": false,
        "disabled": false,
        "expand": false,
        "href": "",
        "index": null,
        "label": "",
        "large": false,
        "loading": false,
        "onClick": [Function],
        "size": 28,
        "style": Object {},
        "tabIndex": -1,
        "tag": "button",
        "type": "button",
      },
      "ref": null,
      "rendered": 21,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
